MvvmCross.Android.Support/V7.RecyclerView/MvxRecyclerAdapter.cs

```diff
@@ -215,54 +191,47 @@ protected virtual int GetItemsSourcePosition(int viewPosition)
 
         protected virtual void SetItemsSource(IEnumerable value)
         {
+            if (Looper.MainLooper != Looper.MyLooper())
+                Mvx.IoCProvider.Resolve<IMvxLog>().Error("ItemsSource property set on a worker thread. This leads to crash in the RecyclerView. It must be set only from the main thread.");
+
             if (ReferenceEquals(_itemsSource, value) && !ReloadOnAllItemsSourceSets)
-            {
                 return;
-            }
 
             _subscription?.Dispose();
             _subscription = null;
 
-            _itemsSource = value;
-
-            if (_itemsSource != null && !(_itemsSource is IList))
+            if (value != null && !(value is IList))
             {
                 MvxBindingLog.Warning("Binding to IEnumerable rather than IList - this can be inefficient, especially for large lists");
             }
 
-            var newObservable = _itemsSource as INotifyCollectionChanged;
-            if (newObservable != null)
-            {
+            if (value is INotifyCollectionChanged newObservable)
                 _subscription = newObservable.WeakSubscribe(OnItemsSourceCollectionChanged);
-            }
 
+            _itemsSource = value;
             NotifyDataSetChanged();
         }
 
         protected virtual void OnItemsSourceCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
         {
+            if(_subscription == null || _itemsSource == null) //Object disposed
+                return;
+
             if (Looper.MainLooper == Looper.MyLooper())
-            {
                 NotifyDataSetChanged(e);
-            }
             else
-            {
-                var h = new Handler(Looper.MainLooper);
-                h.Post(() => NotifyDataSetChanged(e));
-            }
+                Mvx.IoCProvider.Resolve<IMvxLog>().Error("ItemsSource collection content changed on a worker thread. This leads to crash in the RecyclerView as it will not be aware of changes immediatly and may get a deleted item or update an item with a bad item template. All changes must be synchronized on the main thread.");
```

Comment:
Use MvxLog.Instance.

Reply:
MvxLog is internal and not available is this assembly.

Reply:
https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross.Android.Support/V7.RecyclerView/MvxAndroidLog.cs#L11